`Optional` wrapping pattern broken in 1.5.2.Final

I have an issue similar to #2673 in 1.5.2.Final.  Project to reproduce: https://github.com/facboy/mapstruct-bug

```
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /home/xyz/git/mapstruct-bug/src/main/java/org/facboy/mapstruct/OptionalBeanMapper.java:[6,16] error: Internal error in the mapping processor: java.lang.IllegalArgumentException: long     at jdk.compiler/com.sun.tools.javac.model.JavacTypes.getDeclaredType0(JavacTypes.java:272)     at jdk.compiler/com.sun.tools.javac.model.JavacTypes.getDeclaredType(JavacTypes.java:241)      at org.mapstruct.ap.internal.util.AbstractTypeUtilsDecorator.getDeclaredType(AbstractTypeUtilsDecorator.java:119)     at org.mapstruct.ap.internal.model.source.MethodMatcher$GenericAnalyser.resolve(MethodMatcher.java:384)          at org.mapstruct.ap.internal.model.source.MethodMatcher$GenericAnalyser.lineUp(MethodMatcher.java:173)         at org.mapstruct.ap.internal.model.source.MethodMatcher$GenericAnalyser.access$000(MethodMatcher.java:109)     at org.mapstruct.ap.internal.model.source.MethodMatcher.matches(MethodMatcher.java:70)         at org.mapstruct.ap.internal.model.source.SourceMethod.matches(SourceMethod.java:514)          at org.mapstruct.ap.internal.model.source.selector.TypeSelector.lambda$getMatchingParameterBinding$0(TypeSelector.java:164)    at java.base/java.util.ArrayList.removeIf(ArrayList.java:1672)          at java.base/java.util.ArrayList.removeIf(ArrayList.java:1660)         at org.mapstruct.ap.internal.model.source.selector.TypeSelector.getMatchingParameterBinding(TypeSelector.java:163)     at org.mapstruct.ap.internal.model.source.selector.TypeSelector.getMatchingMethods(TypeSelector.java:77)       at org.mapstruct.ap.internal.model.source.selector.MethodSelectors.getMatchingMethods(MethodSelectors.java:67)        at org.mapstruct.ap.internal.processor.creation.MappingResolverImpl$ResolvingAttempt.getBestMatch(MappingResolverImpl.java:485)         at org.mapstruct.ap.internal.processor.creation.MappingResolverImpl$ResolvingAttempt.getTargetAssignment(MappingResolverImpl.java:230)         at org.mapstruct.ap.internal.processor.creation.MappingResolverImpl$ResolvingAttempt.access$100(MappingResolverImpl.java:169)          at org.mapstruct.ap.internal.processor.creation.MappingResolverImpl.getTargetAssignment(MappingResolverImpl.java:145)          at org.mapstruct.ap.internal.model.PropertyMapping$PropertyMappingBuilder.build(PropertyMapping.java:256)      at org.mapstruct.ap.internal.model.BeanMappingMethod$Builder.applyPropertyNameBasedMapping(BeanMappingMethod.java:1412)        at org.mapstruct.ap.internal.model.BeanMappingMethod$Builder.applyPropertyNameBasedMapping(BeanMappingMethod.java:1376)        at org.mapstruct.ap.internal.model.BeanMappingMethod$Builder.build(BeanMappingMethod.java:294)          at org.mapstruct.ap.internal.processor.MapperCreationProcessor.getMappingMethods(MapperCreationProcessor.java:418)     at org.mapstruct.ap.internal.processor.MapperCreationProcessor.getMapper(MapperCreationProcessor.java:181)     at org.mapstruct.ap.internal.processor.MapperCreationProcessor.process(MapperCreationProcessor.java:135)       at org.mapstruct.ap.internal.processor.MapperCreationProcessor.process(MapperCreationProcessor.java:82)        at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:350)        at org.mapstruct.ap.MappingProcessor.processMapperTypeElement(MappingProcessor.java:330)        at org.mapstruct.ap.MappingProcessor.processMapperElements(MappingProcessor.java:279)          at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:174)        at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:1023)          at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:939)     at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1267)      at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1382)   at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1234)       at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:916)    at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:317)   at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:176)  at jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:64)   at jdk.compiler/com.sun.tools.javac.Main.main(Main.java:50)
[INFO] 1 error
```

--- a/processor/src/main/java/org/mapstruct/ap/internal/model/source/MethodMatcher.java
+++ b/processor/src/main/java/org/mapstruct/ap/internal/model/source/MethodMatcher.java
@@ -374,6 +374,8 @@ public class MethodMatcher {
                             // something went wrong
                             return null;
                         }
+                        // Use the boxed equivalent for the type arguments,
+                        // because a primitive type cannot be a type argument
                         typeArgs[i] = matchingType.getBoxedEquivalent().getTypeMirror();
                     }
                     else {

--- a/processor/src/main/java/org/mapstruct/ap/internal/model/source/MethodMatcher.java
+++ b/processor/src/main/java/org/mapstruct/ap/internal/model/source/MethodMatcher.java
@@ -374,7 +374,7 @@ public class MethodMatcher {
                             // something went wrong
                             return null;
                         }
-                        typeArgs[i] = matchingType.getTypeMirror();
+                        typeArgs[i] = matchingType.getBoxedEquivalent().getTypeMirror();
                     }
                     else {
                         // it is not a type var (e.g. Map<String, T> ), String is not a type var

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2925/Issue2925Mapper.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2925/Issue2925Mapper.java
@@ -1,0 +1,23 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2925;
+
+import java.util.Optional;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+@Mapper
+public interface Issue2925Mapper {
+
+    Issue2925Mapper INSTANCE = Mappers.getMapper( Issue2925Mapper.class );
+
+    Target map(Source source);
+
+    static <T> Optional<T> toOptional(T value) {
+        return Optional.ofNullable( value );
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2925/Issue2925Test.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2925/Issue2925Test.java
@@ -1,0 +1,32 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2925;
+
+import org.mapstruct.ap.testutil.IssueKey;
+import org.mapstruct.ap.testutil.ProcessorTest;
+import org.mapstruct.ap.testutil.WithClasses;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Filip Hrisafov
+ */
+@IssueKey("2925")
+@WithClasses({
+    Issue2925Mapper.class,
+    Source.class,
+    Target.class,
+})
+class Issue2925Test {
+
+    @ProcessorTest
+    void shouldUseOptionalWrappingMethod() {
+        Target target = Issue2925Mapper.INSTANCE.map( new Source( 10L ) );
+
+        assertThat( target.getValue() )
+            .hasValue( 10L );
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2925/Source.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2925/Source.java
@@ -1,0 +1,19 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2925;
+
+public class Source {
+
+    private final long value;
+
+    public Source(long value) {
+        this.value = value;
+    }
+
+    public long getValue() {
+        return value;
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2925/Target.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2925/Target.java
@@ -1,0 +1,22 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2925;
+
+import java.util.Optional;
+
+public class Target {
+
+    private Long value;
+
+    public Optional<Long> getValue() {
+        return Optional.ofNullable( value );
+    }
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    public void setValue(Optional<Long> value) {
+        this.value = value.orElse( null );
+    }
+}

